web3/manager.py

```diff
@@ -198,43 +211,91 @@ async def _coro_make_request(
         self.logger.debug(f"Making request. Method: {method}")
         return await request_func(method, params)
 
+    #
+    # formatted_response parses and validates JSON-RPC responses for expected
+    # properties (result or an error) with the expected types.
+    #
+    # Required properties are not strictly enforced to further determine which
+    # exception to raise for specific cases.
+    #
+    # See also: https://www.jsonrpc.org/specification
+    #
     @staticmethod
     def formatted_response(
         response: RPCResponse,
         params: Any,
         error_formatters: Optional[Callable[..., Any]] = None,
         null_result_formatters: Optional[Callable[..., Any]] = None,
     ) -> Any:
-        if "error" in response:
+        # jsonrpc is not enforced (as per the spec) but if present, it must be 2.0
+        if "jsonrpc" in response and response["jsonrpc"] != "2.0":
+            _raise_bad_response_format(
+                response, 'The "jsonrpc" field must be present with a value of "2.0"'
+            )
+
+        # id is not enforced (as per the spec) but if present, it must be a
+        # string or integer
+        # TODO: v7 - enforce id per the spec
+        if "id" in response:
+            response_id = response["id"]
+            # id is always None for errors
+            if (response_id is None and "error" not in response) or (
```

Comment:
I noticed we didnt handle when an error is present but `response_id` is not None. There are integration failures so curious if this is what we want?

Reply:
Oof... I don't know why there's a deviation from the spec but we should change it accordingly. I think we should definitely check for the case where the response `id` is `None` and there is no `error`. But the case you mention, when there is an `error` and the `id` is not `None`, seems like that's fine for me. Makes sense too to show which request gave the error so I'm not against keeping that 👍🏼

Reply:
Cool, I'll take this out!